File: js/creature-notify.js
import { SPECIES, sexSymbol } from './data.js';
import { state } from './state.js';

export function refineDeathCause(c)
{
  if (c.cause && c.cause !== 'exhaustion') return c.cause;
  if (c.age >= (c.genome?.lifespan ?? 999)) return 'old age';
  if (c.hunger <= 0 && c.thirst <= 0) return 'starvation and dehydration';
  if (c.hunger <= 0) return 'starvation';
  if (c.thirst <= 0) return 'dehydration';
  return c.cause || 'exhaustion';
}

export function inferKillerId(c)
{
  if (c.killedById != null) return c.killedById;
  const slot = c.gpuSlot;
  if (typeof slot === 'number' && slot >= 0)
  {
    for (const o of state.creatures)
    {
      if (o.dead || o === c) continue;
      if (o.gpuTargetSlot === slot || o.target === c.id)
      {
        if (o.state === 'hunt' || o.state === 'huntSearch') return o.id;
      }
    }
  }
  const story = c.lifeStory;
  if (story?.events)
  {
    for (let i = story.events.length - 1; i >= 0; i--)
    {
      const ev = story.events[i];
      if (ev.kind === 'preyedOn' && ev.targetId != null) return ev.targetId;
    }
  }
  return null;
}

export function deathCausePhrase(cause)
{
  switch (cause)
  {
    case 'predation': return 'killed by a predator';
    case 'starvation': return 'starved';
    case 'dehydration': return 'died of thirst';
    case 'starvation and dehydration': return 'starved and died of thirst';
    case 'old age': return 'died of old age';
    case 'exhaustion': return 'succumbed to exhaustion';
    case 'meteor': return 'killed by a meteor';
    case 'removed': return 'removed';
    default: return `died (${cause || 'unknown'})`;
  }
}

export function formatBornEvent(c)
{
  const S = SPECIES[c.sp];
  return `${S.emoji} ${S.label} ${sexSymbol(c.sex)} born <b>(gen ${c.gen})</b>`;
}

export function formatMatedEvent(c, partnerId)
{
  const S = SPECIES[c.sp];
  let partnerHtml = '';
  if (partnerId != null)
  {
    const partner = state.creatures.find(x => x.id === partnerId);
    if (partner)
    {
      const PS = SPECIES[partner.sp];
      partnerHtml = ` with ${PS.emoji} ${PS.label} ${sexSymbol(partner.sex)}`;
    }
  }
  return `${S.emoji} ${S.label} ${sexSymbol(c.sex)} mated${partnerHtml}`;
}

export function formatDiedEvent(c)
{
  const S = SPECIES[c.sp];
  const cause = refineDeathCause(c);
  const phrase = deathCausePhrase(cause);
  return `${S.emoji} ${S.label} ${sexSymbol(c.sex)} ${phrase}`;
}

export function eventFocusIds(c, kind)
{
  if (kind === 'died')
  {
    const killerId = inferKillerId(c);
    if (killerId != null) return { focusId: c.id, altFocusId: killerId };
    return { focusId: c.id, altFocusId: null };
  }
  return { focusId: c.id, altFocusId: null };
}

File: godot/WildlandsEcoSim/EcoSim.Core/Sim/CreatureNotify.cs
using EcoSim.Core.Data;

namespace EcoSim.Core.Sim;

public static class CreatureNotify
{
    public static string RefineDeathCause(Creature c)
    {
        if (c.Cause != "exhaustion") return c.Cause;
        if (c.Age >= c.Genome.Lifespan) return "old age";
        if (c.Hunger <= 0 && c.Thirst <= 0) return "starvation and dehydration";
        if (c.Hunger <= 0) return "starvation";
        if (c.Thirst <= 0) return "dehydration";
        return c.Cause;
    }

    public static string DeathCausePhrase(string cause) => cause switch
    {
        "predation" => "was killed by a predator",
        "starvation" => "starved",
        "dehydration" => "died of thirst",
        "starvation and dehydration" => "starved and died of thirst",
        "old age" => "died of old age",
        "exhaustion" => "succumbed to exhaustion",
        "meteor" => "was killed by a meteor",
        "removed" => "was removed",
        _ => $"died ({cause})",
    };

    public static string SexSymbol(string sex) => sex == "male" ? "♂" : "♀";

    public static string FormatFollowedDeathMessage(SpeciesCatalog catalog, Creature c, SimState state)
    {
        var def = catalog.Get(c.Sp);
        string sex = SexSymbol(c.Sex);
        string cause = RefineDeathCause(c);
        int? killerId = InferKillerId(c, state);
        if (killerId != null)
        {
            var killer = state.Creatures.FirstOrDefault(x => x.Id == killerId);
            if (killer != null)
            {
                var kdef = catalog.Get(killer.Sp);
                return $"{def.Emoji} {def.Label} {sex} (#{c.Id}) was preyed on by {kdef.Emoji} {kdef.Label} #{killer.Id}.";
            }
        }

        if (cause == "predation")
        {
            return $"{def.Emoji} {def.Label} {sex} (#{c.Id}) was killed by a predator.";
        }

        return $"{def.Emoji} {def.Label} {sex} (#{c.Id}) {DeathCausePhrase(cause)}.";
    }

    public static int? InferKillerId(Creature prey, SimState state)
    {
        if (prey.KilledById != null) return prey.KilledById;

        foreach (var c in state.Creatures)
        {
            if (c.Dead || c.Target != prey.Id) continue;
            if (c.State is "hunt" or "huntSearch") return c.Id;
        }

        if (prey.LifeStory != null)
        {
            var preyEvent = prey.LifeStory.Events.LastOrDefault(e => e.Kind == "preyedOn");
            if (preyEvent?.Detail != null && int.TryParse(preyEvent.Detail, out int killerId))
            {
                return killerId;
            }
        }

        return null;
    }

    public static string FormatDiedEvent(SpeciesCatalog catalog, Creature c, int? killerId, SimState state)
    {
        string cause = RefineDeathCause(c);
        var def = catalog.Get(c.Sp);
        if (killerId != null)
        {
            var killer = state.Creatures.FirstOrDefault(x => x.Id == killerId);
            if (killer != null)
            {
                var kdef = catalog.Get(killer.Sp);
                return $"Day {state.Day}: {def.Emoji} {def.Label} #{c.Id} was preyed on by {kdef.Emoji} {kdef.Label} #{killer.Id}";
            }
        }

        return $"Day {state.Day}: {def.Emoji} {def.Label} #{c.Id} died ({cause})";
    }

    public static string FormatBornEvent(SpeciesCatalog catalog, Creature c, SimState state)
    {
        var def = catalog.Get(c.Sp);
        string sex = c.Sex == "male" ? "♂" : "♀";
        return $"Day {state.Day}: {def.Emoji} {def.Label} #{c.Id} {sex} was born";
    }
}
